### PR TITLE
Error window

### PR DESCRIPTION
**Pull request by Svelbeard**
Error window for Generic errors and Competitor related errors

### CODE REVIEW

File: src/openloco/ui/WindowManager.cpp [hunk 1]
```diff
@@ -1859,10 +1859,11 @@ namespace openloco::ui::windows
...
         }
 
-        registers regs;
-        regs.bx = (uint16_t)title;
-        regs.dx = (uint16_t)message;
-        call(0x00431A8A, regs);
+        //registers regs;
+        //regs.bx = (uint16_t)title;
+        //regs.dx = (uint16_t)message;
+        //call(0x00431A8A, regs);
```
Comment (Svelbeard) [inline src/openloco/ui/WindowManager.cpp:1865]: Remove commented out code

File: src/openloco/windows/tooltip.cpp [hunk 2]
```diff
@@ -210,15 +211,16 @@ namespace openloco::ui::tooltip
...
-            call(0x00494E33, regs);
+            //// draw_string_centred_raw
+            //registers regs;
+            //regs.cx = ((width + 1) / 2) + x - 1;
+            //regs.dx = y + 1;
+            //regs.al = 0;
+            //regs.bp = _lineBreakCount;
+            //regs.edi = (uint32_t)dpi;
+            //regs.esi = (uint32_t)&_text[0];
+            //call(0x00494E33, regs);
```
Comment (Svelbeard) [inline src/openloco/windows/tooltip.cpp:222]: Remove commented out code

File: src/openloco/openloco.vcxproj [hunk 3]
```diff
@@ -94,6 +94,7 @@
     <ClCompile Include="windows\CompanyWindow.cpp" />
     <ClCompile Include="windows\constructionwnd.cpp" />
     <ClCompile Include="windows\EditKeyboardShortcut.cpp" />
+    <ClCompile Include="windows\error.cpp" />
```
Comment (duncanspumpkin) [inline src/openloco/openloco.vcxproj:97]: For new files they should be in CamelCase

File: src/openloco/windows/error.cpp [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+            make_widget({ 0, 0 }, { 250, 70 }, widget_type::wt_3, 0),
+            make_widget({ 3, 3 }, { 64, 64 }, widget_type::wt_3, 1),
+            widget_end(),
+        };
+    }
+
+    static char* formatErrorString(string_id title, string_id message, char* buffer)
+    {
+        char* ptr = (char*)buffer;
+        ptr[0] = -112;
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:66]: `control_codes::colour_black`

File: src/openloco/windows/error.cpp [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        ptr++;
+
+        if (title != string_ids::null)
+        {
+            auto args = FormatArguments();
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:71]: Can you get args from the calling function. I would like in a future refactor to have to explicitly pass the args into the error window. (I've not even got round to doing that in OpenRCT2 so its a very distant future.)

File: src/openloco/windows/error.cpp [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    {
+        _byte_9C68EC = 0xFF;
+
+        WindowManager::close(WindowType::error);
+
+        char* buffer = _byte_9C64B3;
+
+        buffer = formatErrorString(title, message, buffer);
+
+        if (buffer[0] != _byte_9C64B3[0])
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:103]: this is an address comparison not a value comparison. We are checking to see if formatErrorString actually did anything and there is a string buffer to use.

File: src/openloco/windows/error.cpp [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+            int width = strWidth + 3;
+            int height = (_word_9C66B3 + 1) * 10 + 4;
+
+            int x, y;
+
+            int maxY = ui::height() - height;
+            y = _cursorY + 26; // Normally, we'd display the tooltip 26 lower
+            if (y > maxY)
+                // If y is too large, the tooltip could be forced below the cursor if we'd just clamped y,
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:128]: Use braces `{}`

File: src/openloco/windows/error.cpp [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+            if (!(_suppressErrorSound & (1 << 0)))
+            {
+                int32_t pan = (error->width / 2) + error->x;
+                audio::play_sound(audio::sound_id::error, pan);
+            }
+        }
+    }
+
+    // 0x00431908
+    void open_with_competitor(string_id title, string_id message, uint8_t competitorId)
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:158]: `camelCase`

File: src/openloco/windows/error.cpp [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    {
+        _byte_9C68EC = competitorId;
+
+        WindowManager::close(WindowType::error);
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:162]: Most of this is the same yank out into a common function.

File: src/openloco/windows/error.cpp [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        // 0x00431C05
+        static void draw(ui::window* self, gfx::drawpixelinfo_t* dpi)
+        {
+            uint16_t x = self->x;
+            uint16_t y = self->y;
+            uint16_t width = self->width;
+            uint16_t height = self->height;
+
+            gfx::draw_rect(dpi, x + 1, y + 1, width - 2, height - 2, 0x2000000 | 45);
+            gfx::draw_rect(dpi, x + 1, y + 1, width - 2, height - 2, 0x2000000 | (116 + objectmgr::get<interface_skin_object>()->colour_09));
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:250]: get the interface skin on a separate line there is too much happening here.

File: src/openloco/windows/error.cpp [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+
+using namespace openloco::interop;
+
+namespace openloco::ui::windows::error
+{
+    static loco_global<uint8_t, 0x00508F09> _suppressErrorSound;
+    static loco_global<ui::widget_t[1], 0x00508F1C> _widgets;
+    static loco_global<char[512], 0x009C64B4> _byte_9C64B3;
+    static loco_global<uint16_t, 0x009C66B3> _word_9C66B3;
+    static loco_global<uint8_t, 0x009C68EC> _byte_9C68EC;
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:22]: competitor id only set on a error with competitor

File: src/openloco/graphics/gfx.cpp [hunk 12]
```diff
@@ -798,6 +798,54 @@ namespace openloco::gfx
...
+    }
+
+    uint16_t wrap_string(const char* buffer, uint16_t stringWidth, uint16_t* breakLineCount)
+    {
+        // gfx_wrap_string
+        registers regs;
+        regs.esi = (uintptr_t)buffer;
+        regs.di = stringWidth;
+        call(0x00495301, regs);
+        *breakLineCount = regs.di;
```
Comment (duncanspumpkin) [inline src/openloco/graphics/gfx.cpp:845]: return a std::pair instead of an out param.

File: src/openloco/windows/error.cpp [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        WindowManager::close(WindowType::error);
+
+        char* buffer = _byte_9C64B3;
+
+        auto args = FormatArguments();
+
+        buffer = formatErrorString(title, message, args, buffer);
+
+        if (*buffer != *_byte_9C64B3)
```
Comment (duncanspumpkin) [inline src/openloco/windows/error.cpp:98]: You are still comparing values here. But should be comparing addresses

File: src/openloco/openloco.vcxproj [hunk 14]
```diff
@@ -94,6 +94,7 @@
     <ClCompile Include="windows\CompanyWindow.cpp" />
     <ClCompile Include="windows\constructionwnd.cpp" />
     <ClCompile Include="windows\EditKeyboardShortcut.cpp" />
+    <ClCompile Include="windows\Error.cpp" />
```
Comment (AaronVanGeffen) [inline src/openloco/openloco.vcxproj:97]: This filename was changed to title case, but the filename in the repo is actually in lowercase still.

Comment (Svelbeard) [reply src/openloco/openloco.vcxproj:97]: I just renamed it in VS, should I have done something different?

Comment (AaronVanGeffen) [reply src/openloco/openloco.vcxproj:97]: Windows file systems are case-insensitive, so the change may just not have been picked up by git. In any case, due to this same case-insensitivity, there's no issue building on Windows, either. We'll get around to renaming it, eventually.

**Pull request merged**